lib/js/jsfetch.nim

```diff
@@ -0,0 +1,78 @@
+## - Fetch for the JavaScript target: https://developer.mozilla.org/docs/Web/API/Fetch_API
+from asyncjs import PromiseJs
+from httpcore import HttpMethod
+export PromiseJs, HttpMethod
+
+when not defined(js) and not defined(nimdoc):
+  {.fatal: "Module jsfetch is designed to be used with the JavaScript backend.".}
+when defined(nodejs):
+  {.warning: "By design 'fetch()' is defined on the web browser, but may not be on NodeJS.".}
+
+template method2cstring(metod: HttpMethod): cstring =
+  ## Template that takes an `HttpMethod` and returns an *Uppercase* `cstring`,
+  ## but *only* for the HTTP Methods that are supported by the fetch API.
+  ## High performance and minimal code compared to just `$(HttpMethod)`.
+  assert metod notin {HttpTrace, HttpOptions, HttpConnect}, "HTTP Method not supported by fetch API"
+  case metod
+  of HttpHead:   "HEAD".cstring
+  of HttpGet:    "GET".cstring
+  of HttpPost:   "POST".cstring
+  of HttpPut:    "PUT".cstring
+  of HttpDelete: "DELETE".cstring
+  of HttpPatch:  "PATCH".cstring
+  else:          "GET".cstring
+
+func fetch*(url: cstring): PromiseJs {.importcpp: "fetch(#)".}
+  ## `fetch()` API, Simple GET only (generates `fetch(url)`).
+  ##
+  ## .. code-block:: nim
+  ##   doAssert fetch("https://nim-lang.org") is PromiseJs
+
+func fetch*(url: cstring, httpMethod: static[HttpMethod], body: cstring): PromiseJs {.asmNoStackFrame.} =
+  ## `fetch()` API, for any HTTP Method (static overload for optimization).
+  ##
+  ## .. code-block:: nim
+  ##   doAssert fetch("https://my.api.org", HttpPost, "body") is PromiseJs
+  # See the generated code if you dont understand why this overload exists.
+  assert url.len > 0, "url must not be empty string"
+  const x: cstring = method2cstring(httpMethod)
+  {.emit: """return fetch(`url`, {method: "`x`", body: `body`});""".}
+
+func fetch*(url: cstring, httpMethod: HttpMethod, body: cstring): PromiseJs {.asmNoStackFrame.} =
+  ## .. code-block:: nim
+  ##   let mthd = HttpPost
+  ##   doAssert fetch("https://my.api.org", mthd, "body") is PromiseJs
+  # See the generated code if you dont understand why this overload exists.
+  assert url.len > 0, "url must not be empty string"
+  let x: cstring = method2cstring(httpMethod)
+  {.emit: """return fetch(`url`, {method: `x`, body: `body`});""".}
+
+func fetch*(url: cstring, httpMethod: HttpMethod, body, mode, cache, redirect,
+    credentials, referrerPolicy: cstring, keepalive = false, integrity = "".cstring,
+    referrer = "client".cstring): PromiseJs {.asmNoStackFrame.} =
+  ## `fetch()` API, for any HTTP Method, overload with all arguments supported by fetch API.
+  # For Code Reviewers: Why dont use "Enum" for argument flags instead?.
+  # Because this is *Low-Level* API,a jshttpclient may stand on top,with Enums,Types,etc
```

Comment:
I did look at the code and I still don't understand why you couldn't simply expose a more typesafe API using enums etc (and removing need for a lot of asserts).

you can simply do the conversion from enum member to cstring inside `fetch`, but expose to users a clean API.

Reply:
No problem, lets compare:

```nim
from asyncjs import PromiseJs
from httpcore import HttpMethod

template method2cstring(metod: HttpMethod): cstring =
  case metod
  of HttpHead:   "HEAD".cstring
  of HttpGet:    "GET".cstring
  of HttpPost:   "POST".cstring
  of HttpPut:    "PUT".cstring
  of HttpDelete: "DELETE".cstring
  of HttpPatch:  "PATCH".cstring
  else:          "GET".cstring

func fetch*(url: cstring, httpMethod: static[HttpMethod], body, mode, cache, redirect, credentials, referrerPolicy: cstring, keepalive = false, integrity = "".cstring, referrer = "client".cstring): PromiseJs {.asmNoStackFrame.} =
  const x: cstring = method2cstring(httpMethod)
  {.emit: """return fetch(`url`, {method: "`x`", body: `body`});""".}

discard fetch("https://my.api.org".cstring, HttpPost, "body".cstring)
```
:arrow_down: :arrow_down: :arrow_down: :arrow_down: 
```javascript
/* Generated by the Nim Compiler v1.3.1 */
var framePtr = null;
var excHandler = 0;
var lastJSError = null;
if (typeof Int8Array === 'undefined') Int8Array = Array;
if (typeof Int16Array === 'undefined') Int16Array = Array;
if (typeof Int32Array === 'undefined') Int32Array = Array;
if (typeof Uint8Array === 'undefined') Uint8Array = Array;
if (typeof Uint16Array === 'undefined') Uint16Array = Array;
if (typeof Uint32Array === 'undefined') Uint32Array = Array;
if (typeof Float32Array === 'undefined') Float32Array = Array;
if (typeof Float64Array === 'undefined') Float64Array = Array;
var nim_program_result = 0;
var global_raise_hook_142018 = [null];
var local_raise_hook_142023 = [null];
var out_of_mem_hook_142026 = [null];
var unhandled_exception_hook_142031 = [null];

function fetch_10201036(url_10201039, body_10201041) {
  return fetch(url_10201039, {method: "POST", body: body_10201041});
}
fetch_10201036("https://my.api.org", "body");

```

---

using Enums, etc
```nim
from asyncjs import PromiseJs
from httpcore import HttpMethod

type
  fetchModes* = enum
    fmCors = "cors", fmNoCors = "no-cors", fmSameOrigin = "same-origin"
  fetchCredentials* = enum
    fcInclude = "include", fcSameOrigin = "same-origin", fcOmit = "omit"
  fetchCaches* = enum
    fchDefault = "default", fchNoStore = "no-store", fchReload = "reload", fchNoCache = "no-cache", fchForceCache = "force-cache"
  fetchRedirects* = enum
    frFollow = "follow", frError = "error", frManual = "manual"
  fetchReferrerPolicies* = enum
    frpNoReferrer = "noreferrer", frpNoReferrerWhenDowngrade = "No-Referrer-When-Downgrade",
    frpOrigin = "origin", frpOriginWhenCrossOrigin = "Origin-When-CrossOrigin", frpUnsafeU = "Unsafe"

func fetch*(url: cstring, httpMethod: HttpMethod, body: cstring,
    mode: fetchModes, cache: fetchCaches, redirect: fetchRedirects, credentials: fetchCredentials,
    referrerPolicy: fetchReferrerPolicies, keepalive = false, integrity = "".cstring,
    referrer = "client".cstring): PromiseJs {.asmNoStackFrame.} =
  let a = $httpMethod
  let b = $mode
  let c = $cache
  let d = $redirect
  let e = $credentials
  let f = $referrerPolicy
  {.emit: """return fetch(`url`, {
      method:         `a`,
      body:           `body`,
      integrity:      (`integrity` || undefined),
      referrer:       `referrer`,
      mode:           `b`,
      credentials:    `e`,
      cache:          `c`,
      redirect:       `d`,
      referrerPolicy: `f`,
      keepalive:      `keepalive`
    });
  """.}

discard fetch("https://my.api.org".cstring, HttpPost, "body".cstring, mode = fmNoCors, cache = fchDefault, redirect = frFollow, credentials = fcOmit, referrerPolicy = frpNoReferrer)
```
:arrow_down: :arrow_down: :arrow_down: :arrow_down: 
```javascript
/* Generated by the Nim Compiler v1.3.1 */
var framePtr = null;
var excHandler = 0;
var lastJSError = null;
if (typeof Int8Array === 'undefined') Int8Array = Array;
if (typeof Int16Array === 'undefined') Int16Array = Array;
if (typeof Int32Array === 'undefined') Int32Array = Array;
if (typeof Uint8Array === 'undefined') Uint8Array = Array;
if (typeof Uint16Array === 'undefined') Uint16Array = Array;
if (typeof Uint32Array === 'undefined') Uint32Array = Array;
if (typeof Float32Array === 'undefined') Float32Array = Array;
if (typeof Float64Array === 'undefined') Float64Array = Array;

var NTI10201023 = {size: 0, kind: 14, base: null, node: null, finalizer: null};
var NTI10201017 = {size: 0, kind: 14, base: null, node: null, finalizer: null};
var NTI10201021 = {size: 0, kind: 14, base: null, node: null, finalizer: null};
var NTI10201019 = {size: 0, kind: 14, base: null, node: null, finalizer: null};
var NTI10201015 = {size: 0, kind: 14, base: null, node: null, finalizer: null};
var NTI112 = {size: 0,kind: 28,base: null,node: null,finalizer: null};
var NTI10011023 = {size: 0, kind: 14, base: null, node: null, finalizer: null};
var NNI10011023 = {kind: 2, offset: 0, typ: null, name: null, len: 9, sons: {"0": {kind: 1, offset: 0, typ: NTI10011023, name: "HttpHead", len: 0, sons: null}, 
"1": {kind: 1, offset: 1, typ: NTI10011023, name: "HttpGet", len: 0, sons: null}, 
"2": {kind: 1, offset: 2, typ: NTI10011023, name: "HttpPost", len: 0, sons: null}, 
"3": {kind: 1, offset: 3, typ: NTI10011023, name: "HttpPut", len: 0, sons: null}, 
"4": {kind: 1, offset: 4, typ: NTI10011023, name: "HttpDelete", len: 0, sons: null}, 
"5": {kind: 1, offset: 5, typ: NTI10011023, name: "HttpTrace", len: 0, sons: null}, 
"6": {kind: 1, offset: 6, typ: NTI10011023, name: "HttpOptions", len: 0, sons: null}, 
"7": {kind: 1, offset: 7, typ: NTI10011023, name: "HttpConnect", len: 0, sons: null}, 
"8": {kind: 1, offset: 8, typ: NTI10011023, name: "HttpPatch", len: 0, sons: null}}};
NTI10011023.node = NNI10011023;
var NNI10201015 = {kind: 2, offset: 0, typ: null, name: null, len: 3, sons: {"0": {kind: 1, offset: 0, typ: NTI10201015, name: "cors", len: 0, sons: null}, 
"1": {kind: 1, offset: 1, typ: NTI10201015, name: "no-cors", len: 0, sons: null}, 
"2": {kind: 1, offset: 2, typ: NTI10201015, name: "same-origin", len: 0, sons: null}}};
NTI10201015.node = NNI10201015;
var NNI10201019 = {kind: 2, offset: 0, typ: null, name: null, len: 5, sons: {"0": {kind: 1, offset: 0, typ: NTI10201019, name: "default", len: 0, sons: null}, 
"1": {kind: 1, offset: 1, typ: NTI10201019, name: "no-store", len: 0, sons: null}, 
"2": {kind: 1, offset: 2, typ: NTI10201019, name: "reload", len: 0, sons: null}, 
"3": {kind: 1, offset: 3, typ: NTI10201019, name: "no-cache", len: 0, sons: null}, 
"4": {kind: 1, offset: 4, typ: NTI10201019, name: "force-cache", len: 0, sons: null}}};
NTI10201019.node = NNI10201019;
var NNI10201021 = {kind: 2, offset: 0, typ: null, name: null, len: 3, sons: {"0": {kind: 1, offset: 0, typ: NTI10201021, name: "follow", len: 0, sons: null}, 
"1": {kind: 1, offset: 1, typ: NTI10201021, name: "error", len: 0, sons: null}, 
"2": {kind: 1, offset: 2, typ: NTI10201021, name: "manual", len: 0, sons: null}}};
NTI10201021.node = NNI10201021;
var NNI10201017 = {kind: 2, offset: 0, typ: null, name: null, len: 3, sons: {"0": {kind: 1, offset: 0, typ: NTI10201017, name: "include", len: 0, sons: null}, 
"1": {kind: 1, offset: 1, typ: NTI10201017, name: "same-origin", len: 0, sons: null}, 
"2": {kind: 1, offset: 2, typ: NTI10201017, name: "omit", len: 0, sons: null}}};
NTI10201017.node = NNI10201017;
var NNI10201023 = {kind: 2, offset: 0, typ: null, name: null, len: 5, sons: {"0": {kind: 1, offset: 0, typ: NTI10201023, name: "noreferrer", len: 0, sons: null}, 
"1": {kind: 1, offset: 1, typ: NTI10201023, name: "No-Referrer-When-Downgrade", len: 0, sons: null}, 
"2": {kind: 1, offset: 2, typ: NTI10201023, name: "origin", len: 0, sons: null}, 
"3": {kind: 1, offset: 3, typ: NTI10201023, name: "Origin-When-CrossOrigin", len: 0, sons: null}, 
"4": {kind: 1, offset: 4, typ: NTI10201023, name: "Unsafe", len: 0, sons: null}}};
NTI10201023.node = NNI10201023;

function setConstr() {
        var result = {};
    for (var i = 0; i < arguments.length; ++i) {
      var x = arguments[i];
      if (typeof(x) == "object") {
        for (var j = x[0]; j <= x[1]; ++j) {
          result[j] = true;
        }
      } else {
        result[x] = true;
      }
    }
    return result;
  

  
}
var ConstSet1 = setConstr(17, 16, 4, 18, 27, 19, 23, 22, 21);
function nimCopy(dest_240023, src_240024, ti_240025) {
  var result_245219 = null;

    switch (ti_240025.kind) {
    case 21:
    case 22:
    case 23:
    case 5:
      if (!(is_fat_pointer_235401(ti_240025))) {
      result_245219 = src_240024;
      }
      else {
        result_245219 = [src_240024[0], src_240024[1]];
      }
      
      break;
    case 19:
            if (dest_240023 === null || dest_240023 === undefined) {
        dest_240023 = {};
      }
      else {
        for (var key in dest_240023) { delete dest_240023[key]; }
      }
      for (var key in src_240024) { dest_240023[key] = src_240024[key]; }
      result_245219 = dest_240023;
    
      break;
    case 18:
    case 17:
      if (!((ti_240025.base == null))) {
      result_245219 = nimCopy(dest_240023, src_240024, ti_240025.base);
      }
      else {
      if ((ti_240025.kind == 17)) {
      result_245219 = (dest_240023 === null || dest_240023 === undefined) ? {m_type: ti_240025} : dest_240023;
      }
      else {
        result_245219 = (dest_240023 === null || dest_240023 === undefined) ? {} : dest_240023;
      }
      }
      nimCopyAux(result_245219, src_240024, ti_240025.node);
      break;
    case 24:
    case 4:
    case 27:
    case 16:
            if (src_240024 === null) {
        result_245219 = null;
      }
      else {
        if (dest_240023 === null || dest_240023 === undefined) {
          dest_240023 = new Array(src_240024.length);
        }
        else {
          dest_240023.length = src_240024.length;
        }
        result_245219 = dest_240023;
        for (var i = 0; i < src_240024.length; ++i) {
          result_245219[i] = nimCopy(result_245219[i], src_240024[i], ti_240025.base);
        }
      }
    
      break;
    case 28:
            if (src_240024 !== null) {
        result_245219 = src_240024.slice(0);
      }
    
      break;
    default: 
      result_245219 = src_240024;
      break;
    }

  return result_245219;

}

function cstrToNimstr(c_225079) {
      var ln = c_225079.length;
  var result = new Array(ln);
  var r = 0;
  for (var i = 0; i < ln; ++i) {
    var ch = c_225079.charCodeAt(i);

    if (ch < 128) {
      result[r] = ch;
    }
    else {
      if (ch < 2048) {
        result[r] = (ch >> 6) | 192;
      }
      else {
        if (ch < 55296 || ch >= 57344) {
          result[r] = (ch >> 12) | 224;
        }
        else {
            ++i;
            ch = 65536 + (((ch & 1023) << 10) | (c_225079.charCodeAt(i) & 1023));
            result[r] = (ch >> 18) | 240;
            ++r;
            result[r] = ((ch >> 12) & 63) | 128;
        }
        ++r;
        result[r] = ((ch >> 6) & 63) | 128;
      }
      ++r;
      result[r] = (ch & 63) | 128;
    }
    ++r;
  }
  return result;
  

  
}

function reprEnum(e_315480, typ_315481) {
  var result_315482 = null;

    if (!(is_undefined_315483(typ_315481.node.sons[e_315480]))) {
    result_315482 = nimCopy(null, makeNimstrLit(typ_315481.node.sons[e_315480].name), NTI112);
    }
    else {
      result_315482 = nimCopy(null, (cstrToNimstr((e_315480)+"") || []).concat(makeNimstrLit(" (invalid data!)") || []), NTI112);
    }
    

  return result_315482;

}

var nim_program_result = 0;
var global_raise_hook_142018 = [null];
var local_raise_hook_142023 = [null];
var out_of_mem_hook_142026 = [null];
var unhandled_exception_hook_142031 = [null];
if (!Math.trunc) {
  Math.trunc = function(v) {
    v = +v;
    if (!isFinite(v)) return v;
    return (v - v % 1) || (v < 0 ? -0 : v === 0 ? v : 0);
  };
}


function is_undefined_315483(x_315486) {
  var result_315487 = false;

    result_315487 = x_315486 === undefined;

  return result_315487;

}

function makeNimstrLit(c_225062) {
      var ln = c_225062.length;
  var result = new Array(ln);
  for (var i = 0; i < ln; ++i) {
    result[i] = c_225062.charCodeAt(i);
  }
  return result;
  

  
}

function is_fat_pointer_235401(ti_235403) {
  var result_235404 = false;

  BeforeRet: do {
    result_235404 = !((ConstSet1[ti_235403.base.kind] != undefined));
    break BeforeRet;
  } while (false);

  return result_235404;

}

function nimCopyAux(dest_240028, src_240029, n_240031) {
    switch (n_240031.kind) {
    case 0:
      break;
    case 1:
            dest_240028[n_240031.offset] = nimCopy(dest_240028[n_240031.offset], src_240029[n_240031.offset], n_240031.typ);
    
      break;
    case 2:
          for (var i = 0; i < n_240031.sons.length; i++) {
      nimCopyAux(dest_240028, src_240029, n_240031.sons[i]);
    }
    
      break;
    case 3:
            dest_240028[n_240031.offset] = nimCopy(dest_240028[n_240031.offset], src_240029[n_240031.offset], n_240031.typ);
      for (var i = 0; i < n_240031.sons.length; ++i) {
        nimCopyAux(dest_240028, src_240029, n_240031.sons[i][1]);
      }
    
      break;
    }

  
}

function fetch_10201043(url_10201045, http_method_10201046, body_10201047, mode_10201048, cache_10201049, redirect_10201050, credentials_10201051, referrer_policy_10201052, keepalive_10201053, integrity_10201054, referrer_10201055) {
    var a_10230205 = reprEnum(http_method_10201046, NTI10011023);
    var b_10230210 = reprEnum(mode_10201048, NTI10201015);
    var c_10230215 = reprEnum(cache_10201049, NTI10201019);
    var d_10230220 = reprEnum(redirect_10201050, NTI10201021);
    var e_10230225 = reprEnum(credentials_10201051, NTI10201017);
    var f_10230230 = reprEnum(referrer_policy_10201052, NTI10201023);
    return fetch(url_10201045, {
      method:         a_10230205,
      body:           body_10201047,
      integrity:      (integrity_10201054 || undefined),
      referrer:       referrer_10201055,
      mode:           b_10230210,
      credentials:    e_10230225,
      cache:          c_10230215,
      redirect:       d_10230220,
      referrerPolicy: f_10230230,
      keepalive:      keepalive_10201053
    });
}
fetch_10201043("https://my.api.org", 2, "body", 1, 0, 0, 2, 0, false, "", "client");
```

:slightly_smiling_face:

Reply:
So what, please use enums.

Reply:
@juancarlospaco I don't think it's important to optimize for "less LOC" in the backend, performance is the only thing that truly matters :P

Reply:
@juancarlospaco the fine print here is that in any non toy application using `nim js`, you'll anyway generate a certain amount of glue code, so any saving on code size here will be amount to negligible difference.

Note I mean code size and not performance (which matters a lot more because some application may use a function in a hot loop) because `fetch` and network overhead will dwarf any savings here